### PR TITLE
Issue #13279: Support JDK 20 preview language features

### LINKED ISSUE #13279: Record pattern causes checkstyle to fail in JDK 20 (preview)
I have read check documentation: https://checkstyle.org/config_xxxxxx.html#NameOfAffectedCheck
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```zsh
king@Joses-Mac-mini test % javac --enable-preview --source=20 DeconstructionBug.java
Note: DeconstructionBug.java uses preview features of Java SE 20.
Note: Recompile with -Xlint:preview for details.

king@Joses-Mac-mini test % cat checkstyle.xml                                                                  
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
        <module name="BeforeExecutionExclusionFileFilter">
                <property name="fileNamePattern" value="module\-info\.java$"/>
        </module>
        <module name="SuppressionSingleFilter">
                <property name="files" value="[/\\]target[/\\]"/>
                <property name="checks" value=".*"/>
        </module>
        <module name="TreeWalker">
                <module name="InvalidJavadocPosition"/>
                <module name="JavadocBlockTagLocation"/>
                <module name="JavadocContentLocationCheck"/>
                <module name="JavadocMethod"/>
                <module name="JavadocMissingLeadingAsterisk"/>
                <module name="JavadocMissingWhitespaceAfterAsterisk"/>
                <!--            <module name="JavadocParagraph"/>-->
                <module name="JavadocStyle"/>
                <module name="JavadocTagContinuationIndentation">
                        <property name="offset" value="1"/>
                </module>
                <module name="JavadocType"/>
                <module name="JavadocVariable">
                        <property name="scope" value="protected"/>
                </module>
                <module name="MissingJavadocMethod"/>
                <module name="MissingJavadocType"/>
                <module name="NonEmptyAtclauseDescription"/>
                <module name="RequireEmptyLineBeforeBlockTagGroup"/>
                <module name="SingleLineJavadoc">
                        <property name="ignoreInlineTags" value="false"/>
                </module>
                <!--            <module name="SummaryJavadocCheck"/>-->
                <!--            <module name="WriteTag"/>-->
        </module>
        <!--    <module name="JavadocPackage"/>-->
</module>

king@Joses-Mac-mini test % cat DeconstructionBug.java
import java.util.ArrayList;
import java.util.List;

public class DeconstructionBug
{
        public record InnerRecord(String a, String b)
        {
        }

        public DeconstructionBug()
        {
                List<InnerRecord> aList = new ArrayList<>(1);
                aList.add(new InnerRecord("a", "b"));
                for (InnerRecord(String a, String b) : aList)
                        {
                                System.out.println(a);
                                System.out.println(b);
                        }
        }
}

king@Joses-Mac-mini test % RUN_LOCALE="-Duser.language=en -Duser.country=US"
king@Joses-Mac-mini test % java $RUN_LOCALE -jar ~/Downloads/checkstyle-10.12.0-all.jar -c checkstyle.xml DeconstructionBug.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing DeconstructionBug.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:306)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:223)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: IllegalStateException occurred while parsing file /Users/king/Projects/test/DeconstructionBug.java.
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:105)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:152)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:98)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:334)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:293)
        ... 5 more
Caused by: java.lang.IllegalStateException: 13:11: mismatched input '(' expecting ';'
        at com.puppycrawl.tools.checkstyle.JavaParser$CheckstyleErrorListener.syntaxError(JavaParser.java:255)
        at org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)
        at org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)
        at org.antlr.v4.runtime.DefaultErrorStrategy.reportInputMismatch(DefaultErrorStrategy.java:327)
        at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:139)
        at com.puppycrawl.tools.checkstyle.CheckstyleParserErrorStrategy.recoverInline(CheckstyleParserErrorStrategy.java:38)
        at org.antlr.v4.runtime.Parser.match(Parser.java:208)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.statement(JavaLanguageParser.java:7036)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.blockStatement(JavaLanguageParser.java:6197)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.constructorBlock(JavaLanguageParser.java:5910)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.constructorDeclaration(JavaLanguageParser.java:3068)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.memberDeclaration(JavaLanguageParser.java:2758)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classBodyDeclaration(JavaLanguageParser.java:2670)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classBody(JavaLanguageParser.java:2476)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classDeclaration(JavaLanguageParser.java:1095)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.types(JavaLanguageParser.java:752)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.typeDeclaration(JavaLanguageParser.java:666)
        at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.compilationUnit(JavaLanguageParser.java:413)
        at com.puppycrawl.tools.checkstyle.JavaParser.parse(JavaParser.java:99)
        ... 9 more
Caused by: org.antlr.v4.runtime.InputMismatchException
        ... 23 more
Checkstyle ends with 1 errors.
```

---

Currently using a record pattern deconstruction in a `for` loop causes checkstyle to fail. The error message from checkstyle does not indicate the line where the deconstruction is used, but several lines before it.

Changing the example code from
```java
for (InnerRecord(String a, String b) : aList)
    {
        System.out.println(a);
        System.out.println(b);
    }
```
to one without deconstruction
```java
for (var innerRecord : aList)
    {
        System.out.println(innerRecord.a());
        System.out.println(innerRecord.b());
    }
```
```zsh
king@Joses-Mac-mini test % java $RUN_LOCALE -jar ~/Downloads/checkstyle-10.12.0-all.jar -c checkstyle.xml DeconstructionBug.java
Starting audit...
[ERROR] /Users/king/Projects/test/DeconstructionBug.java:4:1: Missing a Javadoc comment. [MissingJavadocType]
[ERROR] /Users/king/Projects/test/DeconstructionBug.java:6:9: Missing a Javadoc comment. [MissingJavadocType]
[ERROR] /Users/king/Projects/test/DeconstructionBug.java:10:9: Missing a Javadoc comment. [MissingJavadocMethod]
```

### PR DESCRIPTION
**Pull request by nrmancuso**
Closes #13279 

-----------------------

## Check Regression Reports
https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/sevntu-check-regression_part_1-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part4-report/index.html - no dif

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part2-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part3-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part5-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part6-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/part1-report/index.html - no diff

https://nrmancuso.github.io/reports/issue-13279/2023-07-20-T-17-39-42/sevntu-check-regression_part_2-report/index.html - no diff

---------------------
## ANTLR regression reports

https://nrmancuso.github.io/reports/issue-13279/2023-07-21-T-06-50-01/antlr-report/index.html

### CODE REVIEW

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java20/ExpectedJava20RecordDecompositionEnhancedForLoopTricky.txt [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+        |       |       |   |   |   |   `--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [32:36]
+        |       |       |   |   |   |       |--MODIFIERS -> MODIFIERS [32:36]
+        |       |       |   |   |   |       |   `--FINAL -> final [32:36]
+        |       |       |   |   |   |       |--TYPE -> TYPE [32:42]
+        |       |       |   |   |   |       |   `--IDENT -> var [32:42]
+        |       |       |   |   |   |       `--IDENT -> y [32:46]
+        |       |       |   |   |   `--RPAREN -> ) [32:47]
+        |       |       |   |   |--COLON -> : [32:49]
+        |       |       |   |   `--EXPR -> EXPR [32:51]
+        |       |       |   |       `--IDENT -> points1 [32:51]
```
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java20/ExpectedJava20RecordDecompositionEnhancedForLoopTricky.txt:210]: Example of basic record component decomposition in enhanced for loop:
```
 for (Point(final var x, final var y) : points1) {
  }
```

**Review by romani — CHANGES_REQUESTED** [review]
good, only minors:

**Comment by nrmancuso** [comment]
Moving this back to draft, I want to double check some things at https://docs.oracle.com/javase/specs/jls/se20/preview/specs/patterns-switch-record-patterns-jls.html

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java20/InputJava20RecordDecompositionEnhancedForLoopTricky.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+//non-compiled with javac: Compilable with Java20
+package com.puppycrawl.tools.checkstyle.grammar.java20;
+
+import java.util.List;
+
+public class InputJava20RecordDecompositionEnhancedForLoopTricky {
+    void m() {
+
+        // Compiler is failing for annotations on some record components,
+        // ticket is opened with openjdk.
```
[the remark this answers is not in the document]
Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java20/InputJava20RecordDecompositionEnhancedForLoopTricky.java]: I do not have a ticket number, it is still in review. I created an issue for us to track this at https://github.com/checkstyle/checkstyle/issues/13418#issuecomment-1646848517

File: config/jsoref-spellchecker/whitelist.words [hunk 3]
```diff
@@ -289,6 +289,7 @@ dcm
 Dconnection
 Ddry
 declarationorder
+Decomp
```
[the remark this answers is not in the document]
Comment (nrmancuso) [reply config/jsoref-spellchecker/whitelist.words]: Done

**Review by romani — APPROVED** [review]
ok to merge.

**Comment by nrmancuso** [comment]
@romani @rnveach ok, I have resolved my doubts, see conclusion of investigation at https://github.com/checkstyle/checkstyle/pull/13291#issuecomment-1646851409

**Review by rnveach — APPROVED** [review]

**Pull request merged**